File: Java/org/oobench/exceptions/ExceptionsPerformance.java
package org.oobench.exceptions;

import java.util.*;

public class ExceptionsPerformance extends ExceptionsBenchmark {

    public int getMinorNumber() {
        return 2;
    }

    public static void simpleException() throws Exception {
        throw new Exception();
    }

    public static void deepException2() 
            throws IllegalArgumentException, IllegalAccessException {
        throw new IllegalAccessException();
    }

    public static void deepException() {
        try {
            deepException2();
        } catch (IllegalArgumentException iae) { 
        } catch (IllegalAccessException iae) {
        }
    }

    public void except(int count) {
        reset();
        for (int i = 0; i < count; i++) {
            proceed();
            try {
                simpleException();
            } catch (Exception e) {}
        }
        reset();
    }

    public void exceptDeep(int count) {
        reset();
        for (int i = 0; i < count; i++) {
            proceed();
            try {
                deepException();
            } catch (Exception e) {}
        }
        reset();
    }

    public static void main(String[] args) {
        showLocation();
        testExceptions(ExceptionsPerformance.class, 500000, "exceptions", 
                args);
    }
}

File: C++/Exceptions/ExceptionsPerformance.cpp
#include <iostream>
#include "ExceptionsBenchmark.cpp"

namespace oobench {
    struct SyntaxError {
    };
    struct ZeroDivide {
    };
    struct Unexcpected {
    };

    class ExceptionsPerformance : public ExceptionsBenchmark {
    public:
        ExceptionsPerformance(int argc, const char** argv) 
                : ExceptionsBenchmark(argc, argv) { 
        }

        virtual int getMinorNumber() const {
            return 2;
        }

        static void simpleException() {
            throw ZeroDivide();
        }

        static void deepException2() {
            throw Unexcpected();
        }

        static void deepException() {
            try {
                deepException2();
            } catch (SyntaxError& se) {
            } catch (ZeroDivide& zd) {
            }
        }

        virtual void except(unsigned long items) {
            for (unsigned long i = 0; i < items; ++i) {
                try { 
                    simpleException();
                } catch (...) { 
                }
            }
        }

        virtual void exceptDeep(unsigned long items) {
            for (unsigned long i = 0; i < items; ++i) {
                try {
                    deepException();
                } catch (...) {
                }
            }
        }

    };
}

int main(int argc, const char** argv) {
    SHOW_LOCATION("Exceptions");
    try {
        oobench::ExceptionsPerformance performance(argc, argv);
        performance.test(500000L, "exceptions");
    } catch (...) {
        std::cerr << "Unknown exception caught" << std::endl;
    }
    return 0;
    std::exit(0);
}
